Add a diagnostic kstat for obtaining pool status

usaleem-ix opened the pull request:
### Motivation and Context
This PR is an updated version of previous https://github.com/openzfs/zfs/pull/16026

In the original PR, JSON was written into the buffer directly and nvlists were also converted to JSON, which was redundant.

### Description
This PR creates an output nvlist and later nvlist is printed in JSON format to provided buffer. Spares and l2cache devices were also not showing up with previous  https://github.com/openzfs/zfs/pull/16026. This is also fixed now.

### How Has This Been Tested?
Manually tested in different pool configurations.

### Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Performance enhancement (non-breaking change which improves efficiency)
- [ ] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Library ABI change (libzfs, libzfs\_core, libnvpair, libuutil and libzfsbootenv)
- [ ] Documentation (a change to man pages or other documentation)

### Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the OpenZFS [code style requirements](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md#coding-conventions).
- [x] I have updated the documentation accordingly.
- [x] I have read the [**contributing** document](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md).
- [x] I have added [tests](https://github.com/openzfs/zfs/tree/master/tests) to cover my changes.
- [x] I have run the ZFS Test Suite with this change applied.
- [x] All commit messages are properly formatted and contain [`Signed-off-by`](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md#signed-off-by).

usaleem-ix commented on .gitignore:
```diff
@@ -88,3 +88,6 @@ changelog
 *.log
+
+cscope.out
+*,v
```
Removed from this PR.

usaleem-ix commented on include/sys/spa_impl.h:
```diff
@@ -441,6 +442,7 @@ struct spa {
 	uint64_t	spa_autotrim;		/* automatic background trim? */
 	uint64_t	spa_errata;		/* errata issues detected */
 	spa_stats_t	spa_stats;		/* assorted spa statistics */
+	spa_json_stats_t spa_json_stats;	/* diangnostic status in JSON */
```
Updated.

usaleem-ix commented on module/nvpair/nvpair.c:
```diff
@@ -994,6 +994,366 @@ nvlist_remove_nvpair(nvlist_t *nvl, nvpair_t *nvp)
+#define	JPRINTF(start, end, ...)				\
+do {						\
+	if (start < end)	\
+		start += snprintf(start, end - start, __VA_ARGS__);	\
```
Thanks, updated.

usaleem-ix commented on module/zfs/spa.c:
```diff
@@ -9962,6 +9962,7 @@ spa_sync_iterate_to_convergence(spa_t *spa, dmu_tx_t *tx)
 		}
 
 		brt_sync(spa, txg);
+
```
Removed newline

usaleem-ix commented on module/zfs/spa_json_stats.c:
```diff
@@ -0,0 +1,466 @@
+
+	nvlist_add_string(tree, "type", vd->vdev_ops->vdev_op_type);
+
+	/* dRAID vdevs have addtional config keys */
```
Updated, thanks.

tonyhutter commented:
@usaleem-ix is the main goal of this to get the JSON into a kstat so it's lockless? Or was there another use case for wanting the JSON specifically in the kstats? I ask because I'm working on some prototype code that would remove `spa_namespace_lock` from the `zpool status` callpath.  Early tests are promising but I'm not 100% sure it's going to work yet.

usaleem-ix commented:
@tonyhutter yes, the main goal of this is to get `zpool status` JSON into a kstat so it is lockless.

Also, instead of executing `zpool status` command in a subprocess in python, ease of accessing the kstats within our TrueNAS middleware is also one of the motivations for this.

yocalebo commented:
> @usaleem-ix is the main goal of this to get the JSON into a kstat so it's lockless? Or was there another use case for wanting the JSON specifically in the kstats? I ask because I'm working on some prototype code that would remove `spa_namespace_lock` from the `zpool status` callpath. Early tests are promising but I'm not 100% sure it's going to work yet.

@tonyhutter For TrueNAS, we have a few primary reasons for wanting zpool status in procfs.
1. lockless (already mentioned)
2. preventing the necessity for fork+exec'ing (this can become expensive)
3. simplifying a large part of our code-base that uses libzfs (parsing structured output from `procfs` becomes "trivial" compared to using libzfs)

tonyhutter commented:
@usaleem-ix @yocalebo thanks for the info.

1. I just opened https://github.com/openzfs/zfs/pull/16507 for lockless `zpool status`.

2. The kstat means we're doing JSON generation in kernel-space.  That makes me worry more about JSON edge cases (special characters in a vdev name, overflowing the JSON buffer, etc...)

4. This kstat JSON is just a dump of the pool's nvlist.  That means we're exposing a bunch of internal variables that the user never needs to see:

```sh
$ sudo cat /proc/spl/kstat/zfs/tank/status.json | jq
{
  "status_json_version": 4,
  "scl_config_lock": true,
  "scan_error": 2,
  "scan_stats": {
    "func": "NONE",
    "state": "NONE"
  },
...
```
It also API-ifys the config nvlist, which is something I think we should avoid.

5. The `status.json` kstat doesn't match the `zpool status` JSON output.

6. This PR introduces a second JSON implementation in the same codebase.

7. The `zpool status -j` route allows us to use our delegation support, wheras the kstat does not.

Overall, I think putting the JSON functionality in libzfs may be the better route, even if it also means fixing whatever thread safety issues we have in the library.  It's just nice to have all the JSON generation done in userspace.  

Alternatively, if you want to go the fork+exec `zpool status -j` route, it's possible pre-forking could speed things up a bit.

yocalebo commented:
>     1. I just opened [WIP: Remove spa_namespace_lock from zpool status #16507](https://github.com/openzfs/zfs/pull/16507) for lockless `zpool status`.

This is great. Removing that lock for zpool status will benefit everyone (not just selfish developers like myself :smile: )

>     2. The kstat means we're doing JSON generation in kernel-space.  That makes me worry more about JSON edge cases (special characters in a vdev name, overflowing the JSON buffer, etc...)

I tend to agree this isn't that big of a deal. The linux kernel does yaml formatting for certain nfs statistics (and we actually found it wasn't escaping characters properly) and was producing invalid yaml. Not quite the same argument you're making about special chars, over/under flowing etc, but procfs is pretty resilient and has a _ton_ of information from all kinds of esoteric subsystems these days.

>     3. This kstat JSON is just a dump of the pool's nvlist.  That means we're exposing a bunch of internal variables that the user never needs to see:

I agree with you on this point. I see no benefit in exposing this type of information. I was under the assumption this would essentially mirror the `zpool status -j` output (for the most part).

> It also API-ifys the config nvlist, which is something I think we should avoid.

One of the biggest gripes I have is that libzfs isn't versioned and is hard to utilize :smile: I've got no strong opinion on this side of the argument though.
 
>     4. The `status.json` kstat doesn't match the `zpool status` JSON output.

Yeah, this is a problem and I agree with you 100% here. Some subtle differences are okay but not matching at a large percentage is bad.

>     5. This PR introduces a second JSON implementation in the same codebase.

Agree with you 100% on this too. No reason to add unnecessary complexity like this.
 
>     6. The `zpool status -j` route allows us to use our delegation support, wheras the kstat does not.

That's a valid point.

> Overall, I think putting the JSON functionality in libzfs may be the better route, even if it also means fixing whatever thread safety issues we have in the library. It's just nice to have all the JSON generation done in userspace.

Yeah, I'm all for improving libzfs. We've had a couple community members try to user our python libzfs bindings and they have run into issues with non-reentrant calls being made in libzfs (and IIRC, there were some global memory objects wreaking havoc at some point.) Anyways, one of the community members actually tried to help improve thread-safety by opening a PR [here](https://github.com/openzfs/zfs/pull/15793). There were quite a few follow-up commits to make the lib thread-safe. Maybe it is thread-safe (for the most part) and we just need to test it.

> Alternatively, if you want to go the fork+exec `zpool status -j` route, it's possible pre-forking could speed things up a bit.

We already do this with a process pool. However, eventually the child processes get reaped and new ones get forked. This all becomes moot, however, if the library is indeed thread-safe. More tests need to be done on our side I guess.

amotin commented:
@tonyhutter I don't see a big problem in points 2 and 5, since it would be a code not requiring much maintenance, once written and forgotten.  The rest of your points are valid to me.  The question is how much work would be to handle those now and maintain it later.  I see there ~300 lines of `spa_json_stats.c` trying to do it now, and I wonder what more would it take to fix all the JSON output divergence.

amotin commented:
> I'm not convinced the pool kstats are taking the proper locks to deal with device removal/export. When I tested this back in September I was able to panic the kernel by running these two commands in parallel:

I wonder if this issue is not specific to the pool status but is a general bug in Linux kstat implementation, not waiting for the ongoing calls to complete before returning from destruction.

usaleem-ix commented:
I had been working on addressing the review comments, mainly trying to make the kstat output look similar to `zpool status -j`. Output of https://github.com/openzfs/zfs/pull/16026 diverged significantly from what `zpool status -j` outputs and did not output things like checkpoint stats, device removal stats, raidz expansion stats, dedup stats and decoding the error list. Moreover, implementation of status and action strings in `zpool status -j` output will introduce a lot of code duplication.

When kstat output gets large for fairly large zpool, the number of times we construct the output and go back and forth for allocating buffer of suitable size to contain the kstat output, it actually becomes slower than `zpool status -j`.

While all these problems can be looked into and fixed, but this increases the effort required here to address these issues and those that have been highlighted above in earlier comments.

After evaluating the cost and value this brings, we have decided to not pursue this further. Although, anybody who is interested in this work is welcome to try and continue it in future.
